### ISSUE TITLE
Python 3.10 match statement causing false positive unsupported-membership-test

### ISSUE THREAD

**Issue by dragonpaw** [PyCQA/pylint#5967]
### Bug description

```python
"""Strange pylint bug?

This code, when checked with pylint, will report an error on the line 25-27
"""


def test(thing: str | list | None) -> list[str]:
    """Breaks pylint. But not if written as an if statement. Has to be a match."""
    match thing:
        case list():
            return thing
        case str():
            return [thing]
        case None:
            return []
        case _:
            raise ValueError("WTF?")


test("foo")
test(list("foo"))
test(None)

# reports: Value 'test('foo')' doesn't support membership test (unsupported-membership-test)
print("foo" in test("foo"))
print("foo" in test(list("foo")))
print("foo" in test(None))

test(42)  # Raises correctly

```

### Command used

```shell
pylint ~/src/wtf.py
```

### Pylint output

```shell
************* Module wtf
/Users/ash/src/wtf.py:25:15: E1135: Value 'test('foo')' doesn't support membership test (unsupported-membership-test)
/Users/ash/src/wtf.py:26:15: E1135: Value 'test(list('foo'))' doesn't support membership test (unsupported-membership-test)
/Users/ash/src/wtf.py:27:15: E1135: Value 'test(None)' doesn't support membership test (unsupported-membership-test)

--------------------------------------------------------------------
Your code has been rated at -1.54/10 (previous run: -1.54/10, +0.00)
```

### Expected behavior

It should accept that the function will always return a list of strings:

```console
❯ python  ~/src/wtf.py
True
False
False
Traceback (most recent call last):
  File "/Users/ash/src/wtf.py", line 29, in <module>
    test(42)  # Raises correctly
  File "/Users/ash/src/wtf.py", line 17, in test
    raise ValueError("WTF?")
ValueError: WTF?
```

### Pylint version

```shell
pylint 2.12.2
astroid 2.9.3
Python 3.10.3 (main, Mar 21 2022, 18:14:33) [Clang 12.0.5 (clang-1205.0.22.9)]
```

### OS / Environment

Python 3.10.3 via pyenv on a Mac.

**Comment by jacobtylerwalls** [comment]
Thanks for the report. `pylint` relies on `astroid` to infer values of variables, and astroid doesn't understand control-flow from `isinstance()` checks or type annotations, either of which would help solve this problem. Also, it doesn't understand that the catch-all case at the end raises an exception. So `astroid` just infers that `test("foo")` returns `None` (from the function's implicit return).

Related:
control-flow for isinstance, but not in a `match` context: #1498, #1162

**Comment by sinisarudan** [comment]
confirming this too. 
It prevents us from deploying the regularly working 3.10 code with match - case :(

**Comment by sinisarudan** [comment]
is there a workaround to disable this false syntax-error?

**Comment by Pierre-Sassoulas** [comment]
You can use message control (`` # pylint disable=message-name``) see https://pylint.pycqa.org/en/latest/user_guide/messages/message_control.html

**Comment by jacobtylerwalls** [comment]
Duplicate of #5288
